File: app/src/main/java/fi/tuska/jalkametri/util/LogUtil.java
package fi.tuska.jalkametri.util;

import android.util.Log;
import fi.tuska.jalkametri.PrivateData;

public final class LogUtil {

    private LogUtil() {
        // No instantiation required
    }

    /**
     * @return true if logging is enabled and Log.xxx methods should be
     * called; false otherwise.
     */
    public static final boolean isLoggingEnabled() {
        return PrivateData.LOGGING_ENABLED;
    }

    private static String getLogTag(String tag, String message) {
        return "jAlcoMeter";
    }

    private static String getMessage(String tag, String message) {
        return tag + ": " + message;
    }

    /**
     * Wrapper for Log.d; checks whether logging is enabled before calling
     * Log.d, and supports String formatting.
     */
    public static void d(String tag, String message, Object... params) {
        String msg = String.format(message, params);
        if (PrivateData.LOGGING_ENABLED)
            Log.d(getLogTag(tag, msg), getMessage(tag, msg));
    }

    /**
     * Wrapper for Log.i; checks whether logging is enabled before calling
     * Log.i, and supports String formatting.
     */
    public static void i(String tag, String message, Object... params) {
        String msg = String.format(message, params);
        if (PrivateData.LOGGING_ENABLED)
            Log.i(getLogTag(tag, msg), getMessage(tag, msg));
    }

    /**
     * Wrapper for Log.w; checks whether logging is enabled before calling
     * Log.w, and supports String formatting.
     */
    public static void w(String tag, String message, Object... params) {
        String msg = String.format(message, params);
        if (PrivateData.LOGGING_ENABLED)
            Log.w(getLogTag(tag, msg), getMessage(tag, msg));
    }

    /**
     * Wrapper for Log.e; checks whether logging is enabled before calling
     * Log.e, and supports String formatting.
     */
    public static void e(String tag, String message, Object... params) {
        // Error messages are always logged
        String msg = String.format(message, params);
        Log.e(getLogTag(tag, msg), getMessage(tag, msg));
    }

}

File: app/src/main/java/fi/tuska/jalkametri/util/LogUtil.kt
package fi.tuska.jalkametri.util

import android.util.Log
import fi.tuska.jalkametri.Common.LOGGING_ENABLED

interface Logger {
    fun d(tag: String, message: String): Int
    fun i(tag: String, message: String): Int
    fun w(tag: String, message: String): Int
    fun e(tag: String, message: String): Int
}

object AndroidLogger : Logger {
    override fun d(tag: String, message: String) = Log.d(tag, message)
    override fun i(tag: String, message: String) = Log.i(tag, message)
    override fun w(tag: String, message: String) = Log.w(tag, message)
    override fun e(tag: String, message: String) = Log.e(tag, message)
}

object NoopLogger : Logger {
    override fun d(tag: String, message: String) = 0
    override fun i(tag: String, message: String) = 0
    override fun w(tag: String, message: String) = 0
    override fun e(tag: String, message: String) = 0
}

object LogUtil {

    var logger: Logger = AndroidLogger

    val isLoggingEnabled: Boolean = LOGGING_ENABLED

    private fun getLogTag(tag: String, message: String): String = "jAlcoMeter"

    private fun getMessage(tag: String, message: String): String = "$tag: $message"

    fun d(tag: String, message: String, vararg params: Any) {
        val msg = String.format(message, *params)
        if (isLoggingEnabled)
            logger.d(getLogTag(tag, msg), getMessage(tag, msg))
    }

    fun i(tag: String, message: String, vararg params: Any) {
        val msg = String.format(message, *params)
        if (isLoggingEnabled)
            logger.i(getLogTag(tag, msg), getMessage(tag, msg))
    }

    fun w(tag: String, message: String, vararg params: Any) {
        val msg = String.format(message, *params)
        if (isLoggingEnabled)
            logger.w(getLogTag(tag, msg), getMessage(tag, msg))
    }

    fun e(tag: String, message: String, vararg params: Any) {
        // Error messages are always logged
        val msg = String.format(message, *params)
        logger.e(getLogTag(tag, msg), getMessage(tag, msg))
    }

}
